feat(microservice): TCP microservice over TLS

nomaxg opened the pull request:
Builds off of @Flusinerd's work to add the option to connect to a TCP microservice over TLS.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

Adds the ability to pass through TLS options to both the client and the server by means of an optional `tlsOptions` property. 

<!-- Please check the one that applies to this PR using "x". -->
- [ ] Bugfix
- [ x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: https://github.com/nestjs/nest/issues/6745


## What is the new behavior?
If `tlsOptions` are specified, a tls server will be instantiated server-side, and a TLSSocket will be used to connect to the server client-side. Under the hood, options are passed down to the Node [tls library](https://nodejs.org/api/tls.html]). 

## Does this PR introduce a breaking change?
- [ ] Yes
- [ x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Flusinerd commented:
I can confirm, that he sample builds for me as well.

![image](https://user-images.githubusercontent.com/22977969/206785393-7c0d6b65-3596-4498-acf7-cfe2bf9f66bc.png)

Still looking into why it failed on ci, but pretty hard to reproduce

Flusinerd commented on packages/microservices/client/client-tcp.ts:
```diff
@@ -23,13 +25,15 @@ export class ClientTCP extends ClientProxy {
   private readonly socketClass: Type<TcpSocket>;
   private isConnected = false;
   private socket: TcpSocket;
+  public tlsOptions?: ConnectionOptions;
```
You are using ``ConnectionOptions`` here

Flusinerd commented on packages/microservices/interfaces/microservice-configuration.interface.ts:
```diff
@@ -78,6 +79,7 @@ export interface TcpOptions {
     retryAttempts?: number;
     retryDelay?: number;
     serializer?: Serializer;
+    tlsOptions?: TlsOptions;
```
And you are using ``TlsOptions`` here

```suggestion
    tlsOptions?: ConnectionOptions;
```

Flusinerd replied:
And you are using ``TlsOptions`` here

```suggestion
    tlsOptions?: ConnectionOptions;
```

Flusinerd replied:
And you are using ``TlsOptions`` here

```suggestion
    tlsOptions?: ConnectionOptions;
```

Flusinerd commented on packages/microservices/interfaces/microservice-configuration.interface.ts:
```diff
@@ -1,4 +1,5 @@
 import { Type } from '@nestjs/common';
+import { TlsOptions } from 'tls';
```
Update the import
```suggestion
import { ConnectionOptions } from 'tls';
```

Flusinerd reviewed (changes requested):
Seems like you mixed ``ConnectionOptions`` and ``TlsOptions``

nomaxg commented:
> Seems like you mixed `ConnectionOptions` and `TlsOptions`

thanks! updated

nomaxg commented:
> > Seems like you mixed `ConnectionOptions` and `TlsOptions`
> 
> thanks! updated

Actually I believe that the configuration was correct before - `ClientTCP` should use `ConnectionOptions` in the call to [`tls.connect`](https://microsoft.github.io/PowerBI-JavaScript/modules/_node_modules__types_node_tls_d_._tls_.html#connect)

kamilmysliwiec commented:
`samples` workflow is now failing with the following error:

![image](https://user-images.githubusercontent.com/23244943/216037992-ebf98bcf-bcfa-4b7d-a5f4-7abc3402b69e.png)

kamilmysliwiec commented on packages/microservices/client/client-tcp.ts:
```diff
@@ -42,6 +46,10 @@ export class ClientTCP extends ClientProxy {
     this.socket = this.createSocket();
     this.bindEvents(this.socket);
 
+    if (!this.tlsOptions) {
+      this.socket.connect(this.port, this.host);
+    }
```
What was the reason for moving this up? We should also add an inline comment on why "connect" is not required for tls

nomaxg replied:
Moved this back to its original location, added a comment

nomaxg commented:
Thanks for the review, @kamilmysliwiec. Fixed the samples failure by using `ConnectionOptions` instead  of `TlsOptions` in `TcpOptions` (@Flusinerd you were correct about this). Added an inline comment explaining why connecting twice isn't necessary for tls sockets.

kamilmysliwiec commented:
LGTM

kamilmysliwiec commented:
Would you like to create a PR to the docs (with an additional section about this feature) @nomaxg? This could be based off this PR https://github.com/nestjs/docs.nestjs.com/pull/1961
